### PR TITLE
[RAC-5406]: Deliver RackHD vagrant based dev environment code

### PR DESCRIPTION
**Pull request by yaolingling**
**Background**
Set up Vagrant-based demo and development environment for RackHD. For details, go to [https://rackhd.atlassian.net/browse/RAC-5406](url)

**Details**
Use vagrant and ansible to set up rackhd demo and dev environment. Besides, the tutorial pr of how to set up is created. [(https://github.com/RackHD/docs/pull/397)](url)

TestDone
 test it successfully on ubuntu 14.04 and windows.

Reviewer
@iceiilin @pengz1 @anhou

### CODE REVIEW

**Comment by iceiilin** [comment]
add to white list

**Comment by iceiilin** [comment]
test this please

**Comment by iceiilin** [comment]
add to whitelist

**Comment by iceiilin** [comment]
test this please

**Comment by iceiilin** [comment]
add to whitelist

**Comment by iceiilin** [comment]
test this please

File: example/dev-env-setup/Vagrantfile [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+          elsif ENV['REPO'] == "on-core"
+            target.vm.synced_folder "#{ENV['WORKSPACE']}/#{ENV['REPO']}", "/home/vagrant/src/on-taskgraph/node_modules/on-core/"
+            target.vm.synced_folder "#{ENV['WORKSPACE']}/#{ENV['REPO']}", "/home/vagrant/src/on-http/node_modules/on-core/"
+            target.vm.synced_folder "#{ENV['WORKSPACE']}/#{ENV['REPO']}", "/home/vagrant/src/on-tftp/node_modules/on-core/"
+            target.vm.synced_folder "#{ENV['WORKSPACE']}/#{ENV['REPO']}", "/home/vagrant/src/on-dhcp-proxy/node_modules/on-core/"
+          end
+          target.vm.synced_folder "#{ENV['WORKSPACE']}/#{ENV['REPO']}", "/home/vagrant/src/#{ENV['REPO']}"
+          if ENV['CONFIG_DIR']
+            target.vm.synced_folder "#{ENV['WORKSPACE']}/#{ENV['CONFIG_DIR']}", "/opt/onrack/etc/"
+            target.vm.synced_folder "#{ENV['WORKSPACE']}/#{ENV['CONFIG_DIR']}", "/home/vagrant/opt/monorail/"
```
Comment (pengz1) [inline example/dev-env-setup/Vagrantfile]: Should be "/opt/monorail/"

File: example/dev-env-setup/roles/cloneCode/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+   - on-core
+   - on-tasks
+   - on-taskgraph
+   - on-http
+   - on-dhcp-proxy
+   - on-tftp
+   - on-syslog
+
+- name: sync up the latest change of upstream into origin
+  shell: git fetch upstream && git rebase upstream/master
```
Comment (pengz1) [inline example/dev-env-setup/roles/cloneCode/tasks/main.yml]: What is the goal of this operation? Rebase upstream master or to fetch all branches?
git rebase upstream/master should be unnecessary

File: example/dev-env-setup/roles/configNetwork/files/dhcpd.conf [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+deny duplicates;
+
+ignore-client-uids true;
+
+subnet 172.31.128.0 netmask 255.255.252.0 {
+ range 172.31.128.2 172.31.131.254;
+ # Use this option to signal to the PXE client that we are doing proxy DHCP
+ option vendor-class-identifier "PXEClient";
+ option routers 172.31.128.1;
+ option domain-name-servers 10.254.174.10;
```
Comment (pengz1) [inline example/dev-env-setup/roles/configNetwork/files/dhcpd.conf]: Remove this line

File: example/dev-env-setup/roles/config_file/files/config.json [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    "authTokenExpireIn": 86400,
+    "mongo": "mongodb://localhost/pxe",
+    "sharedKey": "qxfO2D3tIJsZACu7UA6Fbw0avowo8r79ALzn+WeuC8M=",
+    "statsd": "127.0.0.1:8125",
+    "syslogBindAddress": "172.31.128.1",
+    "syslogBindPort": 514,
+    "tftpBindAddress": "172.31.128.1",
+    "tftpBindPort": 69,
+    "tftpRoot": "./static/tftp",
+    "minLogLevel": 2
```
Comment (pengz1) [inline example/dev-env-setup/roles/config_file/files/config.json]: For dev enviroment, suggest to change minLogLevel to 0

File: example/dev-env-setup/roles/config_file/files/config.json [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    "dhcpProxyBindAddress": "172.31.128.1",
+    "dhcpProxyBindPort": 4011,
+    "dhcpSubnetMask": "255.255.240.0",
+    "httpEndpoints": [
+        {
+            "address": "0.0.0.0",
+            "port": 8080,
+            "httpsEnabled": false,
+            "proxiesEnabled": true,
+            "authEnabled": true,
```
Comment (pengz1) [inline example/dev-env-setup/roles/config_file/files/config.json]: For dev enviroment, suggest to change authEnable to false

File: example/dev-env-setup/Vagrantfile [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        # NOTE: See further below for instructions on how to use an NFS mount for faster synced file performance
+        if ENV['WORKSPACE']
+          if ENV['REPO'] == "on-tasks"
+            target.vm.synced_folder "#{ENV['WORKSPACE']}/#{ENV['REPO']}", "/home/vagrant/src/on-taskgraph/node_modules/on-tasks/"
+            target.vm.synced_folder "#{ENV['WORKSPACE']}/#{ENV['REPO']}", "/home/vagrant/src/on-http/node_modules/on-tasks/"
+          elsif ENV['REPO'] == "on-core"
+            target.vm.synced_folder "#{ENV['WORKSPACE']}/#{ENV['REPO']}", "/home/vagrant/src/on-taskgraph/node_modules/on-core/"
+            target.vm.synced_folder "#{ENV['WORKSPACE']}/#{ENV['REPO']}", "/home/vagrant/src/on-http/node_modules/on-core/"
+            target.vm.synced_folder "#{ENV['WORKSPACE']}/#{ENV['REPO']}", "/home/vagrant/src/on-tftp/node_modules/on-core/"
+            target.vm.synced_folder "#{ENV['WORKSPACE']}/#{ENV['REPO']}", "/home/vagrant/src/on-dhcp-proxy/node_modules/on-core/"
```
Comment (pengz1) [inline example/dev-env-setup/Vagrantfile]: You should sync on-core first and then on-tasks.
Also I didn't see on-tasks in on-core sync operation.

File: example/dev-env-setup/roles/pm2/tasks/main.yml [hunk 7]
```diff
@@ -0,0 +1,8 @@
+---
+- name: install pm2
+  shell: npm install -g pm2
+
+- name: restart mongodb
```
Comment (pengz1) [inline example/dev-env-setup/roles/pm2/tasks/main.yml]: Why restart mongodb operation is in pm2 action?

File: example/dev-env-setup/roles/snmp/tasks/main.yml [hunk 8]
```diff
@@ -0,0 +1,3 @@
+---
+- name: install snmp
+  apt: pkg=snmp state=installed
```
Comment (pengz1) [inline example/dev-env-setup/roles/snmp/tasks/main.yml]: A general question, I didn't see any version control on tools like snmp, ipmitool, mongo, rabbitmq, isc-dhcp-server and so on.
You should add that.

File: example/dev-env-setup/roles/start-rackhd/files/rackhd.yml [hunk 9]
```diff
@@ -0,0 +1,21 @@
+apps:
```
Comment (pengz1) [inline example/dev-env-setup/roles/start-rackhd/files/rackhd.yml]: You have two copies of rackhd.yml, another one is:
example/dev-env-setup/roles/pm2/files/rackhd.yml

File: example/dev-env-setup/roles/static_files/tasks/main.yml [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+  with_items:
+    - monorail.ipxe
+    - monorail-undionly.kpxe
+    - monorail-efi64-snponly.efi
+    - monorail-efi32-snponly.efi
+    - monorail.intel.ipxe
+  args:
+   chdir: /home/vagrant/src/on-tftp/static/tftp
+
+- name: creare dir /home/vagrant/src/on-http/static/http/common
```
Comment (pengz1) [inline example/dev-env-setup/roles/static_files/tasks/main.yml]: creare => create

File: example/dev-env-setup/Vagrantfile (hunk 6 above)
Comment (yaolingling) [inline example/dev-env-setup/Vagrantfile]: This code block will be removed.

File: example/dev-env-setup/Vagrantfile (hunk 1 above)
Comment (yaolingling) [inline example/dev-env-setup/Vagrantfile]: This code block will be removed.

File: example/dev-env-setup/roles/cloneCode/tasks/main.yml (hunk 2 above)
Comment (yaolingling) [inline example/dev-env-setup/roles/cloneCode/tasks/main.yml]: sync up the latest code change

File: example/dev-env-setup/roles/configNetwork/files/dhcpd.conf (hunk 3 above)
Comment (yaolingling) [inline example/dev-env-setup/roles/configNetwork/files/dhcpd.conf]: configNetwork module will be removed

File: example/dev-env-setup/roles/pm2/tasks/main.yml (hunk 7 above)
Comment (yaolingling) [inline example/dev-env-setup/roles/pm2/tasks/main.yml]: new module for mongodb will be created

File: example/dev-env-setup/roles/snmp/tasks/main.yml (hunk 8 above)
Comment (yaolingling) [inline example/dev-env-setup/roles/snmp/tasks/main.yml]: The third software has already installed in vagrant base box

File: example/dev-env-setup/roles/start-rackhd/files/rackhd.yml (hunk 9 above)
Comment (yaolingling) [inline example/dev-env-setup/roles/start-rackhd/files/rackhd.yml]: The rackhd.yml in /pm2/files will be removed.

File: example/dev-env-setup/roles/static_files/tasks/main.yml (hunk 10 above)
Comment (yaolingling) [inline example/dev-env-setup/roles/static_files/tasks/main.yml]: test this please

**Comment by yaolingling** [comment]
test this please

File: example/dev-env-setup/Vagrantfile [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+        # target.vm.network "private_network", ip: "192.168.50.4"
+
+        # If true, then any SSH connections made will enable agent forwarding.
+        # Default value: false
+        target.ssh.forward_agent = true
+
+        target.vm.provision "shell", inline: <<-SHELL
+          sudo service isc-dhcp-server start
+          
+          sudo su
```
Comment (pengz1) [inline example/dev-env-setup/Vagrantfile]: What is the target for this line?

Comment (yaolingling) [inline example/dev-env-setup/Vagrantfile]: switch to root user

**Pull request merged**